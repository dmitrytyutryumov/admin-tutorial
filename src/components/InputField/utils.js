import { PurchaseUtils } from '../../utils'

const INPUT_TYPES = {
  string: 'text',
  number: 'number',
  boolean: 'checkbox',
  array: 'text',
}

export const getInputType = (value) => {
  const type = typeof value

  return INPUT_TYPES[type] ?? type
}

export const getFieldValue = (value) => {
  let [fieldValue, isCurrencyField] = PurchaseUtils.getCurrencyField(value)

  if (fieldValue instanceof Array) {
    fieldValue = fieldValue[1]
  }
  return [fieldValue, isCurrencyField]
}
