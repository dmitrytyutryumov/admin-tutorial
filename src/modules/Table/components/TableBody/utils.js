import CheckTableCell from './components/CheckTableCell'
import CountryTableCell from './components/CountryTableCell'
import RatingTableCell from './components/RatingTableCell'
import TableCell from './components/TableCell'

const CELLS = {
  country: CountryTableCell,
  rating: RatingTableCell,
  bought: CheckTableCell,
}

export const getTableCell = ({ columns, purchase }) => {
  return columns.map(({ id: columnId }) => {
    const Cell = CELLS[columnId] || TableCell
    return <Cell value={purchase[columnId]} key={columnId} />
  })
}
