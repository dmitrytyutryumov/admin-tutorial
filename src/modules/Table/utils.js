import axios from 'axios'

export const normalizeTable = (columns, users) => {
  const table = columns.map((column) => [column])

  users.forEach((user) => {
    Object.entries(user).forEach((item, idx) => {
      table[idx].push(item[1])
    })
  })
  return table
}

export const sortUsers = ({ users, field, order }) => {
  console.log(users, field, order)
  return [...users].sort((user, nextUser) => {
    let value = user[[field]]
    let nextValue = nextUser[[field]]

    if (value instanceof Array) {
      value = value[1]
      nextValue = nextValue[1]
    }

    if (`${value}`.startsWith('$')) {
      value = parseCurrencyField(value)
      nextValue = parseCurrencyField(nextValue)
    }

    if (typeof value === 'string') {
      value = value.toLowerCase()
      nextValue = nextValue.toLowerCase()
    }

    if (['number', 'boolean'].includes(typeof value)) {
      return (value - nextValue) * order
    }
    return (value > nextValue ? 1 : -1) * order
  })
}

export const parseCurrencyField = (str) => parseFloat(str.replace('$', ''))

export const getUsers = async () => {
  try {
    const response = await axios.get('/purchases')
    return response.data
  } catch {
    return []
  }
}

export const getColumns = async () => {
  try {
    const response = await axios.get('/purchases/meta')
    return response.data
  } catch {
    return []
  }
}
