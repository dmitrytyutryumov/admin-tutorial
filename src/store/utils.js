export const sortPurchases = ({ purchases, field, order }) => {
  return [...purchases].sort((purchase, nextUser) => {
    let value = purchase[[field]]
    let nextValue = nextUser[[field]]

    if (value instanceof Array) {
      value = value[1]
      nextValue = nextValue[1]
    }

    if (`${value}`.startsWith('$')) {
      value = parseCurrencyField(value)
      nextValue = parseCurrencyField(nextValue)
    }

    if (typeof value === 'string') {
      value = value.toLowerCase()
      nextValue = nextValue.toLowerCase()
    }

    if (['number', 'boolean'].includes(typeof value)) {
      return (value - nextValue) * order
    }
    return (value > nextValue ? 1 : -1) * order
  })
}

export const filterPurchases = (purchases, searchQuery) => {
  if (searchQuery === '') {
    return purchases
  }

  return [...purchases].filter((user) => {
    return (
      user['name'].toLowerCase().includes(searchQuery) ||
      user['gameName'].toLowerCase().includes(searchQuery)
    )
  })
}

export const parseCurrencyField = (str) => parseFloat(str.replace('$', ''))
