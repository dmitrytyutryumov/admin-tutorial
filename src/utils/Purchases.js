export const sortPurchases = ({ purchases, field, order }) => {
  if (!field) {
    return [...purchases]
  }
  return [...purchases].sort((purchase, nextUser) => {
    let value = purchase[[field]]
    let nextValue = nextUser[[field]]

    if (value instanceof Array) {
      value = value[1]
      nextValue = nextValue[1]
    }

    if (`${value}`.startsWith('$')) {
      value = parseCurrencyField(value)
      nextValue = parseCurrencyField(nextValue)
    }

    if (typeof value === 'string') {
      value = value.toLowerCase()
      nextValue = nextValue.toLowerCase()
    }

    if (['number', 'boolean'].includes(typeof value)) {
      return (value - nextValue) * order
    }
    return (value > nextValue ? 1 : -1) * order
  })
}

export const filterPurchases = ({ purchases, searchQuery, fields }) => {
  if (searchQuery === '' || !fields) {
    return purchases
  }

  return [...purchases].filter(
    (user) =>
      fields.filter((field) => {
        const fieldValue = user[[field]] || ''
        return fieldValue.toLowerCase().includes(searchQuery)
      }).length !== 0
  )
}

export const parseCurrencyField = (str) => parseFloat(str.replace('$', ''))

export const getCurrencyField = (value) => {
  if (typeof value === 'string' && value.startsWith('$')) {
    return [parseCurrencyField(value), true]
  }
  return [value, false]
}
