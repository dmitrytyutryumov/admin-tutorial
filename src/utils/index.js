export * as PurchaseUtils from './Purchases'

export const capitalize = (str) => str[0].toUpperCase() + str.slice(1)
