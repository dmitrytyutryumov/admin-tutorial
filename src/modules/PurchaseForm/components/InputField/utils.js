import { PurchaseUtils } from '../../../../utils'

export const getInputType = (value) => {
  const valueType = typeof value
  switch (valueType) {
    case 'string':
      return 'text'
    case 'number':
      return 'number'
    case 'boolean':
      return 'checkbox'
    case 'array':
      return 'text'
    default:
      return valueType
  }
}

export const getFieldValue = (value) => {
  let [fieldValue, isCurrencyField] = PurchaseUtils.getCurrencyField(value)

  if (fieldValue instanceof Array) {
    fieldValue = fieldValue[1]
  }
  return [fieldValue, isCurrencyField]
}
