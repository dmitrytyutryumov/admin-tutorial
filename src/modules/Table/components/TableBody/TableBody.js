import React from 'react'
import { useNavigate } from 'react-router-dom'
import { getTableCell } from './utils'

export function TableBody({ purchases, columns }) {
  const navigate = useNavigate()

  const handeClick = (id) => () => navigate(`/purchases/${id}`)

  return (
    <tbody>
      {purchases.map((purchase, id) => (
        <tr key={id} onClick={handeClick(purchase.id)}>
          {getTableCell({ columns: columns, purchase: purchase })}
        </tr>
      ))}
    </tbody>
  )
}
