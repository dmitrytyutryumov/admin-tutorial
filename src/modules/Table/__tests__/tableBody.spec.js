import { render } from '@testing-library/react'
import { TableBody } from '../components/TableBody/TableBody'
import { getTableCell } from '../components/TableBody/utils'

describe('test table cell utils', () => {
  const fakeTable = (cells) => (
    <table>
      <tbody>
        <tr>{cells}</tr>
      </tbody>
    </table>
  )
  it('test getTableCell', () => {
    // Arrange
    const testPurchase = {
      country: 'TestCountry',
      rating: 1,
      bought: '$12',
      phone: '333',
    }
    const testColumns = [
      {
        id: 'country',
      },
      {
        id: 'rating',
      },
      {
        id: 'bought',
      },
      {
        id: 'phone',
      },
    ]
    // Act
    const cells = getTableCell({
      columns: testColumns,
      purchase: testPurchase,
    })
    const { asFragment } = render(fakeTable(cells))
    // Assert
    expect(cells.length).toBe(4)
    expect(asFragment()).toMatchSnapshot()
  })
})

describe('test table body', () => {
  it('test TableBody', () => {
    // Arrange
    const testPurchases = [
      {
        name: 'FakeName',
        phone: '333',
      },
      {
        name: 'FakeName2',
        phone: '222',
      },
    ]
    const testColumns = [
      {
        id: 'name',
      },
      {
        id: 'phone',
      },
    ]
    // Act
    const { asFragment } = render(
      <table>
        <TableBody purchases={testPurchases} columns={testColumns}></TableBody>
      </table>
    )
    // Assert
    expect(asFragment()).toMatchSnapshot()
  })
})
