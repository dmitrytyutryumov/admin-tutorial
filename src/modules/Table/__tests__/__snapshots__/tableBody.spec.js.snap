// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test table body test TableBody 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="table__cell"
        >
          FakeName
        </td>
        <td
          class="table__cell"
        >
          333
        </td>
      </tr>
      <tr>
        <td
          class="table__cell"
        >
          FakeName2
        </td>
        <td
          class="table__cell"
        >
          222
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`test table cell utils test getTableCell 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="table__cell table__country-cell"
        >
          <figure>
            <img
              alt="e icon"
              src="T"
            />
          </figure>
          <div>
            e
          </div>
        </td>
        <td
          class="table__cell"
        >
          <svg>
            star.svg
          </svg>
        </td>
        <td
          class="table__cell"
        >
          <svg>
            check.svg
          </svg>
        </td>
        <td
          class="table__cell"
        >
          333
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
